src/python/pants/jvm/package/war.py

```diff
@@ -0,0 +1,223 @@
+        ),
+        Get(
+            Digest,
+            CreateDigest(
+                [
+                    FileContent(
+                        "make_war.sh",
+                        textwrap.dedent(
+                            f"""\
+                    mkdir -p __war__/WEB-INF/classes
+                    cd __war__
+                    {zip.path} ../output.war -r .
+                    """
+                        ).encode(),
+                        is_executable=True,
+                    )
+                ]
+            ),
+        ),
+    )
+
+    input_digest = await Get(
+        Digest,
+        MergeDigests(
+            [
+                prefixed_jars_digest,
+                descriptor.digest,
+                content.digest,
+                make_archive_script_digest,
+            ]
+        ),
+    )
+
+    result = await Get(
+        ProcessResult,
+        Process(
+            [bash.path, "make_war.sh"],
+            input_digest=input_digest,
+            output_files=("output.war",),
+            description=f"Assemble WAR file for {field_set.address}",
+        ),
+    )
+
+    output_entries = await Get(DigestEntries, Digest, result.output_digest)
+    if len(output_entries) != 1:
+        raise AssertionError("No output from war assembly step.")
+    output_entry = output_entries[0]
+    if not isinstance(output_entry, FileEntry):
+        raise AssertionError("Unexpected digest entry")
+    output_filename = PurePath(field_set.output_path.value_or_default(file_ending="war"))
+    package_digest = await Get(
+        Digest, CreateDigest([FileEntry(str(output_filename), output_entry.file_digest)])
+    )
+    artifact = BuiltPackageArtifact(relpath=str(output_filename))
+    return BuiltPackage(digest=package_digest, artifacts=(artifact,))
+
+
+@rule
+async def render_war_deployment_descriptor(
+    request: RenderWarDeploymentDescriptorRequest,
+) -> RenderedWarDeploymentDescriptor:
+    if not request.descriptor.value:
+        return RenderedWarDeploymentDescriptor(EMPTY_DIGEST)
+
+    descriptor_addresses = await Get(
+        Addresses,
+        UnparsedAddressInputs([request.descriptor.value], owning_address=request.owning_address),
+    )
+    if not descriptor_addresses:
+        return RenderedWarDeploymentDescriptor(EMPTY_DIGEST)
+
+    wrapped_descriptor_target = await Get(WrappedTarget, Address, descriptor_addresses[0])
+    descriptor_target = wrapped_descriptor_target.target
+
+    descriptor_sources = await Get(
+        HydratedSources,
+        HydrateSourcesRequest(
+            descriptor_target[SourcesField],
+            for_sources_types=(FileSourceField, ResourceSourceField),
+            enable_codegen=True,
+        ),
+    )
+
+    descriptor_sources_contents = await Get(
+        DigestContents, Digest, descriptor_sources.snapshot.digest
+    )
+    if len(descriptor_sources_contents) != 1:
+        raise AssertionError("num source != 1")
```

Comment:
Even though you don't expect this to be hit, probably worth more context.

Reply:
Fixed.